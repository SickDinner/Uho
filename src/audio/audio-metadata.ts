import type { AudioAsset } from '@core/audio-assets.ts';

export type AudioCategory = AudioAsset['category'];

export interface AudioCategorySummary {
  category: AudioCategory;
  total: number;
  loaded: number;
}

export function summarizeAudioCategories(
  assets: readonly AudioAsset[],
  loadedIds: ReadonlySet<string> | readonly string[]
): AudioCategorySummary[] {
  const loadedSet = loadedIds instanceof Set ? loadedIds : new Set(loadedIds);
  const categoryMap = new Map<AudioCategory, { total: number; loaded: number }>();

  for (const asset of assets) {
    const entry = categoryMap.get(asset.category) ?? { total: 0, loaded: 0 };
    entry.total += 1;
    if (loadedSet.has(asset.id)) {
      entry.loaded += 1;
    }
    categoryMap.set(asset.category, entry);
  }

  return Array.from(categoryMap.entries()).map(([category, counts]) => ({
    category,
    total: counts.total,
    loaded: counts.loaded,
  }));
}

export const AUDIO_CATEGORIES: AudioCategory[] = [
  'music',
  'ambient',
  'sfx',
  'weather',
  'ui',
  'voice',
];
