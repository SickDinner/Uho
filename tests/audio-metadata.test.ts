import { describe, expect, it } from 'vitest';

import type { AudioAsset } from '@/audio/audio-asset-registry.ts';
import { AUDIO_CATEGORIES, summarizeAudioCategories } from '@/audio/audio-metadata.ts';

describe('summarizeAudioCategories', () => {
  const sampleAssets: AudioAsset[] = [
    { id: 'city_theme', src: 'music', category: 'music', description: 'Music' },
    { id: 'menu_theme', src: 'music', category: 'music', description: 'Music' },
    { id: 'rain_light', src: 'weather', category: 'weather', description: 'Rain' },
    { id: 'ui_click', src: 'ui', category: 'ui', description: 'Click' },
  ];

  it('counts assets per category', () => {
    const summary = summarizeAudioCategories(sampleAssets, new Set(['city_theme', 'ui_click']));
    expect(summary).toEqual([
      { category: 'music', total: 2, loaded: 1 },
      { category: 'weather', total: 1, loaded: 0 },
      { category: 'ui', total: 1, loaded: 1 },
    ]);
  });

  it('accepts array input for loaded ids', () => {
    const summary = summarizeAudioCategories(sampleAssets, ['menu_theme']);
    expect(summary.find(entry => entry.category === 'music')?.loaded).toBe(1);
  });
});

describe('AUDIO_CATEGORIES constant', () => {
  it('lists known categories in a stable order', () => {
    expect(AUDIO_CATEGORIES).toMatchSnapshot();
  });
});
