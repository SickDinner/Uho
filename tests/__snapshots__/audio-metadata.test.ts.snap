// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AUDIO_CATEGORIES constant > lists known categories in a stable order 1`] = `
[
  "music",
  "ambient",
  "sfx",
  "weather",
  "ui",
  "voice",
]
`;
